core/dbt/helper_types.py

```diff
@@ -88,6 +90,63 @@ class NoValue(dbtClassMixin):
     novalue: NVEnum = field(default_factory=lambda: NVEnum.novalue)
 
 
+@dataclass
+class IncludeExclude(dbtClassMixin):
+    INCLUDE_ALL = ("all", "*")
+
+    include: Union[str, List[str]]
+    exclude: List[str] = field(default_factory=list)
+
+    def __post_init__(self):
+        if isinstance(self.include, str) and self.include not in self.INCLUDE_ALL:
+            raise ValidationException(
+                f"include must be one of {self.INCLUDE_ALL} or a list of strings"
+            )
+
+        if self.exclude and self.include not in self.INCLUDE_ALL:
+            raise ValidationException(
+                f"exclude can only be specified if include is one of {self.INCLUDE_ALL}"
+            )
+
+        if isinstance(self.include, list):
+            self._validate_items(self.include)
+
+        if isinstance(self.exclude, list):
+            self._validate_items(self.exclude)
+
+    def includes(self, item_name: str):
+        return (
+            item_name in self.include or self.include in self.INCLUDE_ALL
+        ) and item_name not in self.exclude
+
+    def _validate_items(self, items: List[str]):
+        pass
+
+
+class WarnErrorOptions(IncludeExclude):
+    # TODO: this method can be removed once the click CLI is in use
+    @classmethod
+    def from_yaml_string(cls, warn_error_options_str: Optional[str]):
+
+        # TODO: resolve circular import
+        from dbt.config.utils import parse_cli_yaml_string
+
+        warn_error_options_str = warn_error_options_str or "{}"
+        warn_error_options = parse_cli_yaml_string(warn_error_options_str, "warn-error-options")
+        return cls(
+            include=warn_error_options.get("include", []),
+            exclude=warn_error_options.get("exclude", []),
+        )
+
+    def _validate_items(self, items: List[str]):
+        valid_exception_names = set(
+            [name for name, cls in dbt_event_types.__dict__.items() if isinstance(cls, type)]
+        )
+        for item in items:
+            if item not in valid_exception_names:
+                raise ValidationException(f"{item} is not a valid dbt error name.")
```

Comment:
It seems like env var values might not preserve casing:
```bash
$ DBT_WARN_ERROR_OPTIONS='{"include": ["NoNodesForSelectionCriteria"]}' dbt run --select alsiuhfasd
21:07:25  Running with dbt=1.4.0-b1
21:07:25  Found 3 models, 0 tests, 0 snapshots, 0 analyses, 291 macros, 0 operations, 0 seed files, 0 sources, 0 exposures, 0 metrics
21:07:25  Encountered an error:
Runtime Error
  nonodesforselectioncriteria is not a valid dbt error name.
```
Could we just make this a case-insensitive comparison?

Reply:
> It seems like env var values might not preserve casing:

Yeah, looks like we lower any env var values here: https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/flags.py#L214-L218. 

Does this make sense to do for the [`_NON_BOOLEAN_FLAGS`](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/flags.py#L50-L57) though? For example, lowercasing `LOG_PATH` or `PROFILES_DIR` could lead to surprising (log path different from what  was configured) or non-functional (profiles dir should be case-sensitive) behaviour. Would also be different behaviour depending on whether the flag was configured from the environment vs arg/user config.

Reply:
I take it you're asking, should we be doing that lowercasing at all? I agree it's an odd choice. Specifically the bit about functionally different behavior if passed as env var vs. CLI flag does not feel right! Shall we rip it out? No time like a new minor version to do it

Reply:
done!